packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -147,19 +151,69 @@ class SegmentedControl<T> extends StatefulWidget {
   _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
 }
 
-class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+class _SegmentedControlState<T> extends State<SegmentedControl<T>>
+    with TickerProviderStateMixin<SegmentedControl<T>> {
   T _pressedKey;
 
-  void _onTapDown(T currentKey) {
-    setState(() {
-      _pressedKey = currentKey;
-    });
+  final List<AnimationController> _selectionControllers = <AnimationController>[];
+  final List<ColorTween> _childTweens = <ColorTween>[];
+
+  static final ColorTween forwardBackgroundColorTween = new ColorTween(
+    begin: _kPressedBackground,
+    end: CupertinoColors.activeBlue,
+  );
+
+  static final ColorTween reverseBackgroundColorTween = new ColorTween(
+    begin: CupertinoColors.white,
+    end: CupertinoColors.activeBlue,
+  );
+
+  static final ColorTween textColorTween = new ColorTween(
+    begin: CupertinoColors.activeBlue,
+    end: CupertinoColors.white,
+  );
+
+  @override
+  void initState() {
+    super.initState();
+    setupAnimationControllers();
   }
 
-  void _onTapUp(TapUpDetails event) {
-    setState(() {
-      _pressedKey = null;
-    });
+  @override
+  void dispose() {
+    for (AnimationController animationController in _selectionControllers) {
+      animationController.dispose();
+    }
+    super.dispose();
+  }
+
+  void setupAnimationControllers() {
+    for (T key in widget.children.keys) {
+      final AnimationController animationController = new AnimationController(
+        duration: _kFadeDuration,
+        vsync: this,
+      )..addListener(() {
+          setState(() {
+            // State of background/text colors has changed
+          });
+        });
+
+      if (widget.groupValue == key) {
+        _childTweens.add(reverseBackgroundColorTween);
+        animationController.forward();
+      } else {
+        _childTweens.add(forwardBackgroundColorTween);
+      }
+      _selectionControllers.add(animationController);
+    }
+  }
```

Comment:
Will this cause animations to stall if the widget rebuilds in the middle of an animation? (See "stall" comment below.)